cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/RollingLocationLogAppender.java

```diff
@@ -0,0 +1,229 @@
+                 + getName());
+      return;
+    }
+
+    // logic from OutputStreamAppender
+    if (this.encoder == null) {
+      addStatus(new ErrorStatus("No encoder set for the appender named \""
+                                  + name + "\".", this));
+      return;
+    }
+
+    // TODO remove if condition after removing Guice injection
+    if (locationFactory != null) {
+      this.locationManager = new LocationManager(locationFactory);
+      started = true;
+    }
+  }
+
+  @Override
+  public void doAppend(ILoggingEvent eventObject) throws LogbackException {
+    try {
+      // TODO Remove 'if' after removing Guice injection
+      if (locationFactory != null && this.locationManager == null) {
+        this.locationManager = new LocationManager(locationFactory);
+        started = true;
+      }
+
+      // logic from AppenderBase
+      if (!this.started) {
+        addStatus(new WarnStatus(
+          "Attempted to append to non started appender [" + name + "].",
+          this));
+        return;
+      }
+
+      // logic from AppenderBase
+      if (getFilterChainDecision(eventObject) == FilterReply.DENY) {
+        return;
+      }
+
+      String namespaceId = eventObject.getMDCPropertyMap().get(LocationManager.TAG_NAMESPACE_ID);
+      if (!namespaceId.equals(NamespaceId.SYSTEM.getEntityName())) {
+        LocationIdentifier logLocationIdentifier = locationManager.getLocationIdentifier(eventObject
+                                                                                           .getMDCPropertyMap());
+        rollover(logLocationIdentifier, eventObject);
+        OutputStream locationOutputStream = locationManager.getLocationOutputStream(logLocationIdentifier, fileName);
+        setOutputStream(locationOutputStream);
+        writeOut(eventObject);
+      }
+    } catch (IllegalArgumentException iae) {
+      // this shouldn't happen
+      LOG.error("Unrecognized context ", iae);
+    } catch (IOException ioe) {
+      throw new LogbackException("Exception during append", ioe);
+    }
+  }
+
+  private void rollover(LocationIdentifier logLocationIdentifier, ILoggingEvent eventObject) {
+    if (!locationManager.getActiveFilesLocations().containsKey(logLocationIdentifier)) {
+      return;
+    }
+
+    Location logFile = locationManager.getActiveFilesLocations().get(logLocationIdentifier);
+    triggeringPolicy.setLocation(logFile);
+
+    // Since TriggeringPolicy#isTriggeringEvent method takes File in the api, we provide dummy file to our
+    // triggeringPolicy implementation
+    if (triggeringPolicy.isTriggeringEvent(new File("dummy"), eventObject)) {
+      try {
+        rollingPolicy.setLocation(locationManager.getLogLocation(logLocationIdentifier), logFile);
+        rollingPolicy.rollover();
+
+        // clear caches so new location gets created
+        locationManager.closeOutputStream(locationManager.getActiveFiles().get(logLocationIdentifier));
```

Comment:
closing of output stream should happen before rollover. also its better to use the pattern of having LogOutputStream and that handle the case of closing output stream and evicting entry from map, etc. you can see `CDAPLogAppender` for reference.